mlflow/pipelines/steps/predict.py

```diff
@@ -0,0 +1,153 @@
+            scored_df.reset_index(drop=True),
+            "Profile of Scored Dataset",
+        )
+
+        # Build card
+        card = BaseCard(self.pipeline_name, self.name)
+        # Tab #1: data profile for scored data:
+        card.add_tab("Scored Data Profile", "{{PROFILE}}").add_pandas_profile(
+            "PROFILE", scored_data_profile
+        )
+        # Tab #2: run summary.
+        (
+            card.add_tab(
+                "Run Summary",
+                """
+                {{ SCORED_DATA_NUM_ROWS }}
+                {{ EXE_DURATION}}
+                {{ LAST_UPDATE_TIME }}
+                """,
+            ).add_markdown(
+                "SCORED_DATA_NUM_ROWS",
+                f"**Number of scored dataset rows:** `{len(scored_df)}`",
+            )
+        )
+
+        return card
+
+    def _run(self, output_directory):
+        from pyspark.sql.functions import struct
+
+        run_start_time = time.time()
+
+        # Get or create spark session
+        try:
+            spark = _get_active_spark_session()
+        except Exception as e:
+            raise MlflowException(
+                message=(
+                    "Encountered an error while searching for an active Spark session to"
+                    " score dataset with spark UDF. Please create a Spark session and try again."
+                ),
+                error_code=BAD_REQUEST,
+            ) from e
+        if not spark:
+            raise MlflowException(
+                message=(
+                    "No active SparkSession detected to score dataset with spark UDF. "
+                    "Please create a Spark session and try again."
+                ),
+                error_code=BAD_REQUEST,
+            )
+
+        # read cleaned dataset
+        ingested_data_path = get_step_output_path(
+            pipeline_root_path=self.pipeline_root,
+            step_name="ingest_scoring",
+            relative_path=_INPUT_FILE_NAME,
+        )
+        input_sdf = spark.read.parquet(ingested_data_path)
+
+        # score dataset
+        model_uri = self.step_config["model_uri"]
+        env_manager = "local" if "_disable_env_restoration" in self.step_config else "conda"
+        predict = mlflow.pyfunc.spark_udf(spark, model_uri, env_manager=env_manager)
+        scored_sdf = input_sdf.withColumn("prediction", predict(struct(*input_sdf.columns)))
+
+        # save predictions
+        # note: the current output writing logic allows no overwrites
+        output_format = self.step_config["output_format"]
+        if output_format == "parquet" or output_format == "delta":
+            scored_sdf.coalesce(1).write.format(output_format).save(
+                self.step_config["output_location"]
+            )
+        else:
+            scored_sdf.write.format("delta").saveAsTable(self.step_config["output_location"])
+
+        # predict step artifacts
+        scored_pdf = scored_sdf.toPandas()
```

Comment:
Can't Spark write to Parquet directly? Why convert to Pandas? If this is to build step card, you need to make sure it's small enough. Downsampling should be done here before the conversion.

Reply:
cc @jerrylian-db

Reply:
@jinzhang21 I decided to write the parquet file with Pandas because I needed to convert the Dataframe to pandas anyways for building the step card. Also, the ingest and split steps both write to parquet using pandas, so I wanted to follow the existing pattern.

For downsampling for the step card data profiling, we use the `get_pandas_data_profile` function, which contains downsampling logic. Should we do additional downsampling when writing the step output artifact? Without downsampling, the predict step output artifact would have the same number of rows as the ingest step output artifact.

Reply:
OK with converting to Pandas after downsampling.

On downsampling, if it's not done before conversion, it may not fit within one machine, which invalidates the whole point of having Spark in both of the ingest and predict step.

Reply:
What technique should I use for downsampling? Also, right now both the `DeltaTableDataset` and the `SparkSqlDataset` directly convert spark dataframes to pandas dataframes without downsampling.

Reply:
You can downsample here (just for profiling and step card usage) with Spark Dataframe: https://spark.apache.org/docs/3.1.2/api/python/reference/api/pyspark.sql.DataFrame.sample.html#pyspark-sql-dataframe-sample

Reply:
Spark Dataframe can sample directly like: https://spark.apache.org/docs/3.1.2/api/python/reference/api/pyspark.sql.DataFrame.sample.html#pyspark-sql-dataframe-sample

Added some comments to the PR.

Reply:
Thanks! I've added a max data profiling size of 10000 rows. Choose this limit as it is the same size as the spark dataframe limit in [model evaluation](https://github.com/mlflow/mlflow/blob/4c58179509e6f6047789efb0a95c2b0e20cb6c8f/mlflow/models/evaluation/base.py#L233:5).